content/articles/2019-08-callr-task-q.Rmd

```diff
@@ -0,0 +1,586 @@
+```
+
+The columns are:
+
+* `id`: a character id, which can be user-supplied or auto-assigned (if the
+  user did not supply it). This is useful to identify tasks.
+* `idle`: a logical flag, whether this is a dummy idle task or not.
+* `state`: current state of the task. More about this shortly.
+* `fun`: the function the task needs to run. This is a list column.
+* `args`: arguments to pass to the function. This is a list itself,
+  so the column is a list column.
+* `worker`: the `callr::r_session` object, the R session that is running
+  the task, or `NULL` if the task is not running.
+* `result`: another list column, the result of the run, if the task is
+  already done, `NULL` otherwise.
+
+The possible task states are: _waiting_, _running_, _ready_ and _done_.
+The first two are not very surprising. The distinction between the
+last two is somewhat technical. A task is _ready_ if the background R
+session has finished running it. The queue hasn't read out its result yet,
+and the R session is still assigned to it. (I.e. the task's `worker` column
+is not `NULL`.) A task is _done_ if the queue has already read out the
+result of the function call, and has reassigned the R session to another
+task, so its `worker` column is `NULL`.
+
+Tasks that are _running_ and _ready_ always have an R session assigned to
+them. Since all R sessions are always assigned to tasks (dummy idle tasks,
+if there is nothing else), this means the that the sum of the _running_
+and _ready_ tasks always equals the number of workers.
+
+The idle tasks are somewhat special, because they are never _done_. If
+an idle task is _ready_ and its worker is reassigned, it will be _waiting_
+again. They are also almost never in the _running_ state. When the queue
+assigns a worker to an idle task, the task will immediately go into the
+_ready_ state, since the queue is immediately allowed to re-assign the
+worker, should a real task be _waiting_.
+
+However, when a worker is starting up, its idle task is _running_, until
+their background R process has started up. After this initial _running_
+state the idle tasks are always either _waiting_ or _ready_.
+
+## Implementation
+
+I am ready to start the implementation now. I'll focus on the individual
+methods here, and show the complete code of the R6 class at the end.
+Let's start with the internal data. `tasks` contains the task data frame,
+`initialize()` will create it. `next_id` and `get_next_id` will provide
+us unique task ids. I prefix these with a dot, to increase the probability
+that they won't interfere with user supplied task ids. So they'll
+be `".1"`, `".2"`, etc.
+
+```r
+  private = list(
+    tasks = NULL,
+    next_id = 1L,
+    get_next_id = function() {
+      id <- private$next_id
+      private$next_id <- id + 1L
+      paste0(".", id)
+    }
+  )
+```
+
+The `initialize()` method will just defer the work to a private method and
+return the object itself. Returning `self` is usually good practice and
```

Comment:
I think you could drop the discussion of `invisible(self)` here; it's a bit of a distraction.